src/abstractdataframe/reshape.jl

```diff
@@ -399,3 +399,109 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+
+Note: The eltypes of columns in resulting `DataFrame`
+(other than the first column, which always has eltype `String`)
+will depend on the eltypes of _all_ input columns
+based on the results of `prote_type`.
+That is, if the source `DataFrame` contains `String` and `Int` columns,
+resulting columns will have eltype `Any`.
+If the source has a mix of numeric types (eg. `Float64` and `Int`),
+columns in resulting `DataFrame` will be promoted to `Float64`.
+
+# Examples
+
+```jldoctest
+julia> df1 = DataFrame(a=["x", "y"], b=[1.,2.], c=[3,4], d=[true,false])
+2×4 DataFrame
+│ Row │ a      │ b       │ c     │ d    │
+│     │ String │ Float64 │ Int64 │ Bool │
+├─────┼────────┼─────────┼───────┼──────┤
+│ 1   │ x      │ 1.0     │ 3     │ 1    │
+│ 2   │ y      │ 2.0     │ 4     │ 0    │
+
+julia> df2 = DataFrame(a=["x", "y"], b=[1, "two"], c=[3,4], d=[true,false])
+2×4 DataFrame
+│ Row │ a      │ b   │ c     │ d    │
+│     │ String │ Any │ Int64 │ Bool │
+├─────┼────────┼─────┼───────┼──────┤
+│ 1   │ x      │ 1   │ 3     │ 1    │
+│ 2   │ y      │ two │ 4     │ 0    │
+
+julia> permutedims(df1) # note the column types
+3×3 DataFrame
+│ Row │ a      │ x       │ y       │
+│     │ String │ Float64 │ Float64 │
+├─────┼────────┼─────────┼─────────┤
+│ 1   │ b      │ 1.0     │ 2.0     │
+│ 2   │ c      │ 3.0     │ 4.0     │
+│ 3   │ d      │ 1.0     │ 0.0     │
+
+julia> permutedims(df2)
+3×3 DataFrame
+│ Row │ a      │ x   │ y   │
+│     │ String │ Any │ Any │
+├─────┼────────┼─────┼─────┤
+│ 1   │ b      │ 1   │ two │
+│ 2   │ c      │ 3   │ 4   │
+│ 3   │ d      │ 1   │ 0   │
+```
+"""
+function Base.permutedims(df::AbstractDataFrame, src_namescol::ColumnIndex,
+                          dest_namescol::Union{Symbol, AbstractString};
+                          makeunique::Bool=false)
+
+    nrow(df) > 0 || throw(
+        ArgumentError("`permutedims` not defined for data frame with 0 rows"))
+    eltype(df[!, src_namescol]) <: SymbolOrString || throw(
+            ArgumentError("src_namescol must have eltype `Symbol` or `<:AbstractString`"))
+
+    df_notsrc = df[!, Not(src_namescol)]
+    df_permuted = DataFrame(dest_namescol => names(df_notsrc))
+
+    m = permutedims(Matrix(df_notsrc))
+    df_tmp = rename!(DataFrame(Tables.table(m)), df[!, src_namescol], makeunique=makeunique)
+    return hcat!(df_permuted, df_tmp, copycols=false)
+end
+
+function Base.permutedims(df::AbstractDataFrame, src_namescol::ColumnIndex;
+                          makeunique::Bool=false)
+    if src_namescol isa Integer
+        1 <= src_namescol <= ncol(df) || throw(ArgumentError("`src_namescol` doesn't exist"))
+        dest_namescol = _names(df)[src_namescol]
+    else
+        dest_namescol = src_namescol
+    end
+    return permutedims(df, src_namescol, dest_namescol; makeunique=makeunique)
+end
+
+Base.permutedims(df::AbstractDataFrame; makeunique::Bool=false) =
+    permutedims(df, 1; makeunique=makeunique)
```

Comment:
Why not just use `src_namescol::ColumnIndex=1` in the last definition?

Reply:
I swear there was a reason... either it was wrong or doesn't apply anymore